### PR TITLE
dump-c: fix output of _Bool-typed constants

### PR DESCRIPTION
**Pull request by tautschnig**
Only __CPROVER_bool-typed constants can be used with exprt::is_true.
exprt::is_one's documentation promised that ID_c_bool was supported,
which wasn't actually the case. This commit fixes this and makes use of
is_one to fix the bug in dump-c.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by kroening — APPROVED** [review]

File: regression/goto-instrument/dump-bool/test.desc [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+--dump-c
+VERIFICATION SUCCESSFUL
+^EXIT=0$
+^SIGNAL=0$
+--
+^warning: ignoring
+irep
+--
+Dump-c previously always generated 0 for _Bool or __CPROVER_bool-typed
+constants. This test ensures that the value "1" is correctly generated.
```
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-instrument/dump-bool/test.desc]: I don't understand how this test tests this?

File: src/goto-instrument/dump_c.cpp [hunk 2]
```diff
@@ -1521,22 +1521,48 @@ void dump_ct::cleanup_type(typet &type)
     !type.get(ID_tag).empty());
 }
 
+static expr2c_configurationt expr2c_configuration()
+{
+  static expr2c_configurationt configuration{
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-instrument/dump_c.cpp:1535]: IMHO would be better written as a series of assignments, then you don't need the comments.

File: src/goto-instrument/dump_c_class.h [hunk 3]
```diff
@@ -156,7 +152,7 @@ class dump_ct
   symbol_tablet copied_symbol_table;
   const namespacet ns;
   const dump_c_configurationt dump_c_config;
-  std::unique_ptr<languaget> language;
+  const irep_idt mode;
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-instrument/dump_c_class.h:155]: This change should be a separate commit with an explanation for why it's done. It's not obvious to me that this is an improvement, and in any case it doesn't seem to be directly related to fixing anything with the output of booleans? Not saying I'm against this, just that the commit should do a better job of explaining it.

File: regression/goto-instrument/dump-bool/test.desc (hunk 1 above)
Comment (martin-cs) [reply regression/goto-instrument/dump-bool/test.desc]: Yeah... I'm not seeing it either.  Not quite sure where the `irep` is coming from.

File: src/goto-instrument/dump_c.cpp [hunk 4]
```diff
@@ -526,7 +526,7 @@ void dump_ct::convert_compound(
       if(t.get_width()<=config.ansi_c.long_long_int_width)
         struct_body << "long long int " << comp_name
           << " : " << t.get_width();
-      else if(language->id()=="cpp")
+      else if(mode == ID_cpp)
```
Comment (martin-cs) [inline src/goto-instrument/dump_c.cpp:529]: Doesn't seem to correspond with the PR message.  I'm fine with them going in as a minor fix but just wanted to flag it in case it was a mistake.

**Review by martin-cs — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: regression/goto-instrument/dump-bool/test.desc (hunk 1 above)
Comment (tautschnig) [reply regression/goto-instrument/dump-bool/test.desc]: Test description expanded: the previous (buggy) output made dump-c produce v = 0 instead of v = 1.

File: src/goto-instrument/dump_c_class.h (hunk 3 above)
Comment (tautschnig) [reply src/goto-instrument/dump_c_class.h:155]: Broken out into a second commit.
